File: Python/Best_Time_to_Buy_and_Sell_Stock.py

'''
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/

Companies: Adobe, Alation, Amazon, Apple, Atlassian, Bloomberg, ByteDance, Capital One, Cisco, Docusign, Expedia, Goldman Sachs, JP Morgan, Netflix, Orcale, Paypal, ServiceNow, Uber, Visa, Walmart, Zoho, Zoom, eBay, tcs
prices = [7, 1, 5, 3, 6, 4] #Return 5 as maxProfit
'''

def maxProfit(prices: list) -> int:
    minPrice = prices[0]
    maxProfit = 0
    for i in range(0, len(prices)): 
        if prices[i] < minPrice:
            minPrice = prices[i]
        
        maxProfit = max(maxProfit, prices[i] - minPrice)
    return maxProfit

prices = [7, 1, 5, 3, 6, 4] #Return 5 as maxProfit
print(maxProfit(prices))

File: JavaScript/best_time_to_buy_and_sell_stock.js
function maxProfit(prices){
    let maxProfit = 0
    let minPrice = prices[0]

    for (let sell = 1; sell < prices.length; sell++){
        let sellPrice = prices[sell]
        let profit = sellPrice - minPrice

        maxProfit = Math.max(maxProfit, profit)

        if(sellPrice < minPrice) minPrice = sellPrice
    }
    return maxProfit
}

prices = [7, 1, 5, 3, 6, 4]
console.log(maxProfit(prices));